### ISSUE TITLE
Type of cls in implementation of asbtact classmethod is not checked on declaration

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#1697]
**Describe the bug**
When implementing an abstract `@classmethod` for child class `Child`, one can incorrectly type `cls` as `Child` (as opposed to `Type[Child]`), and get no "overrides method in incompatible manner" error.

**To Reproduce**
Run pyright on the following code:

```python
from typing import TypeVar, Type
from abc import abstractmethod, ABC

# pyright: strict


SELF = TypeVar("SELF", bound="Parent")
class Parent(ABC):
    @abstractmethod
    @classmethod
    def from_string(cls: Type[SELF], value: str) -> SELF:
        pass

class Child(Parent):
    def instance_method(self):
        print("do instance method")

    @classmethod
    def from_string(cls: "Child", value: str) -> "Child":
        cls.instance_method() #incorrectly using an instance method
        return Child()

child = Child() # no error when instatiating Child class

# Directly using the overridden method gives a "could not bind method" error
# but it's far away from the actual source of the issue.
Child.from_string("blas") # pyright error

# Using the method indirectly produces no compiler errors, though
Parent.__subclasses__()[0].from_string("blas") # no pyright error

```

**Expected behavior**
An error on the line that implements Child.from_string. `mypy` for examples shows this:

`error: The erased type of self "Child" is not a supertype of its class "Type[Child]"`

**VS Code extension or command-line**
Pyright extension version 1.1.123, command  line version 1.1.126

**Comment by erictraut** [comment]
Thanks for the suggestion. This will be implemented in the next version of pyright and pylance.
